[win] slow IO.getBasedFile

since [JDK ](https://bugs.openjdk.org/browse/JDK-8207005) disable the file canonicalization cache File.getCanonicalPath is very slow on windows.
example sampling during building eclipse platform workspace.
![image](https://github.com/bndtools/bnd/assets/51790620/8f12c9dc-4287-47e5-97d7-aa533e73dc97)

the zipslip check in https://github.com/bndtools/bnd/blame/f6371af9346e75f0170987ceb52ce40cf5a8fa9b/aQute.libg/src/aQute/lib/io/IO.java#L850
could be much faster to just return child in the normal case that the filename does not contain an directory escape:
`if (!file.contains("..")) return getFile(base, file);`
which avoid costly file name capitalization at all.

your JUnit with you implemenatation fails on windows:
```
java.lang.AssertionError: 
Expecting actual:
  "java.io.IOException: The relative path 'bar/../../..' when normalized starts with ..: '..\..'"
to contain:
  "The relative path 'bar/../../..' when normalized starts with ..: '../..'" 
	testGetBasedFile(IO.java:53)
```

--- a/aQute.libg/src/aQute/lib/io/IO.java
+++ b/aQute.libg/src/aQute/lib/io/IO.java
@@ -77,9 +77,7 @@ import aQute.lib.stringrover.StringRover;
 import aQute.libg.glob.Glob;
 
 public class IO {
-	private static final boolean						isWindows				= File.separatorChar == '\\';
 	final static EnvironmentCalculator					hc						= new EnvironmentCalculator();
-	final static Function<String, String>				getenv;
 	private static final Pattern						WINDOWS_MACROS			= Pattern.compile("%(?<key>[^%]+)%");
 	private static final int							BUFFER_SIZE				= IOConstants.PAGE_SIZE * 16;
 	private static final int							DIRECT_MAP_THRESHOLD	= BUFFER_SIZE;
@@ -90,13 +88,50 @@ public class IO {
 	private static final EnumSet<StandardOpenOption>	writeOptions			= EnumSet.of(StandardOpenOption.WRITE,
 		StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
 	private static final EnumSet<StandardOpenOption>	readOptions				= EnumSet.of(StandardOpenOption.READ);
+	final static Path									DOTDOT					= Path.of("..");
+
+	interface OS {
+		/**
+		 * Resolve the subPath from the absolute file base. The subPath must not
+		 * resolve to a path that is not in the subtree of files from base. it
+		 * may contain .. as long as this does not end up at the start when
+		 * normalized. It (surprisingly) can handle absolute files since `new
+		 * File(base,"/foo") resolves not to "/foo" ... I.e. except for the ..,
+		 * it can never escape base
+		 *
+		 * @param base the absolute base directory
+		 * @param subPath the sub path
+		 * @return the absolute file resolved from base with relativePath
+		 * @throws IOException
+		 */
+
+		File getBasedFile(File base, String subPath) throws IOException;
+
+		/**
+		 * Get an environment variable. Windows is special ...
+		 *
+		 * @param variableName the variable name
+		 * @return the env variable or null
+		 */
+		String getenv(String variableName);
+
+		/**
+		 * Convert to a safe name on the current platform. This is the _name_
+		 * part only, not a path!
+		 *
+		 * @param name the file name
+		 * @return a safe file name
+		 */
+		String toSafeFileName(String name);
+	}
+
+	final static OS os = File.separatorChar == '\\' ? new Windows() : new Other();
 
 	static {
-		getenv = isWindows ? hc::getenv : System::getenv;
-		home = hc.getDirLocation(getenv.apply("HOME"), System.getProperty("user.home"));
+		home = hc.getDirLocation(os.getenv("HOME"), System.getProperty("user.home"));
 		String javaHome = System.getProperty("java.home")
 			.replaceAll("(/|\\\\)jre$", "");
-		JAVA_HOME = hc.getDirLocation(getenv.apply("JAVA_HOME"), javaHome);
+		JAVA_HOME = hc.getDirLocation(os.getenv("JAVA_HOME"), javaHome);
 	}
 
 	public static String getExtension(String fileName, String deflt) {
@@ -657,7 +692,8 @@ public class IO {
 	public static ByteBuffer read(Path path) throws IOException {
 		try (FileChannel in = readChannel(path)) {
 			long size = in.size();
-			if (!isWindows && (size > DIRECT_MAP_THRESHOLD)) {
+
+			if (!isWindows() && (size > DIRECT_MAP_THRESHOLD)) {
 				return in.map(MapMode.READ_ONLY, 0, size);
 			}
 			ByteBuffer bb = ByteBuffer.allocate((int) size);
@@ -847,15 +883,21 @@ public class IO {
 		return f.getAbsoluteFile();
 	}
 
-	public static File getBasedFile(File base, String file) throws IOException {
-		base = base.getCanonicalFile();
-		File child = getFile(base, file);
-		if (child.getCanonicalPath()
-			.startsWith(base.getCanonicalPath()
-				.concat(File.separator))) {
-			return child;
-		}
-		throw new IOException("The file " + child + " is outside of the base " + base);
+	/**
+	 * Resolve the subPath from the absolute file base. The subPath must not
+	 * resolve to a path that is not in the subtree of files from base. it may
+	 * contain .. as long as this does not end up at the start when normalized.
+	 * It (surprisingly) can handle absolute files since `new File(base,"/foo")
+	 * resolves not to "/foo" ... I.e. except for the .., it can never escape
+	 * base
+	 *
+	 * @param base the absolute base directory
+	 * @param subPath the sub path
+	 * @return the absolute file resolved from base with relativePath
+	 * @throws IOException
+	 */
+	public static File getBasedFile(File base, String subPath) throws IOException {
+		return os.getBasedFile(base, subPath);
 	}
 
 	public static Path getPath(String file) {
@@ -1526,7 +1568,7 @@ public class IO {
 	 */
 	public static boolean createSymbolicLinkOrCopy(Path link, Path target) {
 		try {
-			if (isWindows || !createSymbolicLink(link, target)) {
+			if (isWindows() || !createSymbolicLink(link, target)) {
 				// only copy if target length and timestamp differ
 				BasicFileAttributes targetAttrs = Files.readAttributes(target, BasicFileAttributes.class);
 				try {
@@ -1606,36 +1648,11 @@ public class IO {
 		};
 
 	public static String toSafeFileName(String string) {
-		StringBuilder sb = new StringBuilder();
-
-		for (int i = 0; i < string.length(); i++) {
-			char c = string.charAt(i);
-			if (c < ' ')
-				continue;
-
-			if (isWindows) {
-				sb.append(switch (c) {
-					case '<', '>', '"', '/', '\\', '|', '*', ':' -> '%';
-					default -> c;
-				});
-			} else {
-				sb.append(switch (c) {
-					case '/', ':' -> '%';
-					default -> c;
-				});
-			}
-		}
-		if (sb.length() == 0 || (isWindows && RESERVED_WINDOWS_P.matcher(sb)
-			.matches()))
-			sb.append("_");
-
-		return sb.toString();
+		return os.toSafeFileName(string);
 	}
 
-	private final static Pattern RESERVED_WINDOWS_P = Pattern.compile("CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9]");
-
 	public static boolean isWindows() {
-		return isWindows;
+		return os instanceof Windows;
 	}
 
 	/*
@@ -1705,7 +1722,7 @@ public class IO {
 	 *         null if not found
 	 */
 	public static String getenv(String key) {
-		return getenv.apply(key);
+		return os.getenv(key);
 	}
 
 	public static String readUTF(DataInput in) throws IOException {

--- a/aQute.libg/src/aQute/lib/io/IO.java
+++ b/aQute.libg/src/aQute/lib/io/IO.java
@@ -889,14 +889,19 @@ public class IO {
 	 * contain .. as long as this does not end up at the start when normalized.
 	 * It (surprisingly) can handle absolute files since `new File(base,"/foo")
 	 * resolves not to "/foo" ... I.e. except for the .., it can never escape
-	 * base
+	 * base.
+	 * <p>
+	 * If the subPath is null or empty, this will return the base
 	 *
 	 * @param base the absolute base directory
-	 * @param subPath the sub path
+	 * @param subPath the sub path or null
 	 * @return the absolute file resolved from base with relativePath
 	 * @throws IOException
 	 */
 	public static File getBasedFile(File base, String subPath) throws IOException {
+		assert base != null : "base must not be null or the result could be anywhere on the fs";
+		if (subPath == null || subPath.isEmpty())
+			return base;
 		return os.getBasedFile(base, subPath);
 	}
 

--- a/aQute.libg/src/aQute/lib/io/Other.java
+++ b/aQute.libg/src/aQute/lib/io/Other.java
@@ -1,0 +1,49 @@
+package aQute.lib.io;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+
+import aQute.lib.io.IO.OS;
+
+class Other implements OS {
+
+	@Override
+	public File getBasedFile(File base, String subPath) throws IOException {
+		String use;
+		if (subPath.contains("..")) {
+			Path normalizedPath = Path.of(subPath)
+				.normalize();
+			if (normalizedPath.getNameCount() > 0 && normalizedPath.getName(0)
+				.equals(IO.DOTDOT)) {
+				throw new IOException("io.sub.up invalid path, will escape the designated directory. path='" + subPath
+					+ "', base='" + base + "', normalized='" + normalizedPath + "'");
+			}
+			use = normalizedPath.toString();
+		} else
+			use = subPath;
+		return new File(base, use);
+	}
+
+	@Override
+	public String getenv(String string) {
+		return System.getenv(string);
+	}
+
+	@Override
+	public String toSafeFileName(String string) {
+		StringBuilder sb = new StringBuilder();
+
+		for (int i = 0; i < string.length(); i++) {
+			char c = string.charAt(i);
+			if (c < ' ')
+				continue;
+
+			sb.append(switch (c) {
+				case '/', ':' -> '%';
+				default -> c;
+			});
+		}
+		return sb.toString();
+	}
+}

--- a/aQute.libg/src/aQute/lib/io/Windows.java
+++ b/aQute.libg/src/aQute/lib/io/Windows.java
@@ -1,0 +1,70 @@
+package aQute.lib.io;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import aQute.lib.io.IO.OS;
+
+class Windows implements OS {
+	final static Pattern WINDOWS_BAD_FILE_NAME_P = Pattern.compile(
+		"(?:(:?.*[\u0000-\u001F<>:\"|/\\\\?*].*)|..|CON|PRN|AUX|NUL|COM\\d|COM¹|COM²|COM³|LPT\\d|LPT¹|LPT²|LPT³)(?:\\.\\w+)?",
+		Pattern.CASE_INSENSITIVE);
+
+	@Override
+	public File getBasedFile(File base, String subPath) throws IOException {
+		String use;
+		Matcher matcher = WINDOWS_BAD_FILE_NAME_P.matcher(subPath);
+		if (matcher.find()) {
+
+			Path normalizedPath = Path.of(subPath)
+				.normalize();
+
+			if (normalizedPath.startsWith(IO.DOTDOT)) {
+				throw new IOException("io.sub.up invalid path, will escape the designated directory. path='" + subPath
+					+ "', base='" + base + "', normalized='" + normalizedPath + "'");
+			}
+			for (int i = 0; i < normalizedPath.getNameCount(); i++) {
+				String segment = normalizedPath.getName(i)
+					.toString();
+				if (matcher.reset(segment)
+					.matches()) {
+					throw new IOException("io.win.sub.invalid pathcontains reserved names on windows. path='" + subPath
+						+ "', base='" + base + "', pattern='" + WINDOWS_BAD_FILE_NAME_P + "'");
+				}
+			}
+			use = normalizedPath.toString();
+		} else
+			use = subPath;
+		return new File(base, use);
+	}
+
+	@Override
+	public String getenv(String string) {
+		return IO.hc.getenv(string);
+	}
+
+	@Override
+	public String toSafeFileName(String string) {
+		StringBuilder sb = new StringBuilder();
+
+		for (int i = 0; i < string.length(); i++) {
+			char c = string.charAt(i);
+			if (c < ' ')
+				continue;
+
+			sb.append(switch (c) {
+				case '<', '>', '"', '/', '\\', '|', '*', ':', '?' -> '%';
+				default -> c;
+			});
+		}
+		if (sb.length() == 0 || WINDOWS_BAD_FILE_NAME_P.matcher(sb)
+			.matches())
+			sb.append("_");
+
+		return sb.toString();
+	}
+
+}

--- a/aQute.libg/src/aQute/lib/io/Windows.java
+++ b/aQute.libg/src/aQute/lib/io/Windows.java
@@ -10,7 +10,7 @@ import aQute.lib.io.IO.OS;
 
 class Windows implements OS {
 	final static Pattern WINDOWS_BAD_FILE_NAME_P = Pattern.compile(
-		"(?:(:?.*[\u0000-\u001F<>:\"|/\\\\?*].*)|..|CON|PRN|AUX|NUL|COM\\d|COM¹|COM²|COM³|LPT\\d|LPT¹|LPT²|LPT³)(?:\\.\\w+)?",
+		"(?:(:?.*[\u0000-\u001F<>:\"|/\\\\?*].*)|\\.\\.|CON|PRN|AUX|NUL|COM\\d|COM¹|COM²|COM³|LPT\\d|LPT¹|LPT²|LPT³)(?:\\.\\w+)?",
 		Pattern.CASE_INSENSITIVE);
 
 	@Override

--- a/aQute.libg/src/aQute/lib/io/Windows.java
+++ b/aQute.libg/src/aQute/lib/io/Windows.java
@@ -19,23 +19,28 @@ class Windows implements OS {
 		Matcher matcher = WINDOWS_BAD_FILE_NAME_P.matcher(subPath);
 		if (matcher.find()) {
 
-			Path normalizedPath = Path.of(subPath)
-				.normalize();
+			try {
+				Path normalizedPath = Path.of(subPath)
+					.normalize();
 
-			if (normalizedPath.startsWith(IO.DOTDOT)) {
-				throw new IOException("io.sub.up invalid path, will escape the designated directory. path='" + subPath
-					+ "', base='" + base + "', normalized='" + normalizedPath + "'");
-			}
-			for (int i = 0; i < normalizedPath.getNameCount(); i++) {
-				String segment = normalizedPath.getName(i)
-					.toString();
-				if (matcher.reset(segment)
-					.matches()) {
-					throw new IOException("io.win.sub.invalid pathcontains reserved names on windows. path='" + subPath
-						+ "', base='" + base + "', pattern='" + WINDOWS_BAD_FILE_NAME_P + "'");
+				if (normalizedPath.startsWith(IO.DOTDOT)) {
+					throw new IOException("io.sub.up invalid path, will escape the designated directory. path='"
+						+ subPath + "', base='" + base + "', normalized='" + normalizedPath + "'");
 				}
+				for (int i = 0; i < normalizedPath.getNameCount(); i++) {
+					String segment = normalizedPath.getName(i)
+						.toString();
+					if (matcher.reset(segment)
+						.matches()) {
+						throw new IOException("io.win.sub.invalid pathcontains reserved names on windows. path='"
+							+ subPath + "', base='" + base + "', pattern='" + WINDOWS_BAD_FILE_NAME_P + "'");
+					}
+				}
+				use = normalizedPath.toString();
+			} catch (java.nio.file.InvalidPathException e) {
+				throw new IOException("io.win.sub.invalid pathcontains reserved names on windows. path='" + subPath
+					+ "', base='" + base + "': '" + e.getMessage() + "'");
 			}
-			use = normalizedPath.toString();
 		} else
 			use = subPath;
 		return new File(base, use);

--- a/aQute.libg/src/aQute/lib/io/package-info.java
+++ b/aQute.libg/src/aQute/lib/io/package-info.java
@@ -1,4 +1,4 @@
-@Version("4.4.0")
+@Version("4.5.0")
 package aQute.lib.io;
 
 import org.osgi.annotation.versioning.Version;

--- a/aQute.libg/test/aQute/lib/io/IOTest.java
+++ b/aQute.libg/test/aQute/lib/io/IOTest.java
@@ -558,6 +558,8 @@ public class IOTest {
 		assertThat("foo/💩../bar".contains("..")).isTrue();
 		assertThat("foo/.💩./bar".contains("..")).isFalse();
 		File base = new File("base").getAbsoluteFile();
+		assertThat(new File(base, "/foo").getName()).isEqualTo("foo");
+		assertThat(new File(base, "/foo").getParentFile()).isEqualTo(base);
 		assertThat(new File(base, "")).isEqualTo(base);
 		assertThat(new File(base, "./").getParentFile()).isEqualTo(base);
 		assertThat(new File(base, "/bar").getParentFile()).isEqualTo(base);
@@ -568,12 +570,32 @@ public class IOTest {
 		assertThat(IO.getBasedFile(base, ".💩.")).isEqualTo(new File(base, ".💩."));
 		assertThat(except(() -> IO.getBasedFile(base, ".."))).contains("io.sub.up");
 		assertThat(except(() -> IO.getBasedFile(base, "bar/../../.."))).contains("io.sub.up");
+		assertThat(new File(base, "/foo").getName()).isEqualTo("foo");
+		assertThat(new File(base, "/foo").getParentFile()).isEqualTo(base);
 	}
 
+	@EnabledOnOs(WINDOWS)
 	@Test
-	public void testCoder() {
-		String a = "💩..";
-		"..".indexOf(a);
+	public void testOnWindowsOnly() {
+		File base = new File("base").getAbsoluteFile();
+		assertThat(new File(base, "c:\\foo").getName()).isEqualTo("foo");
+		assertThat(new File(base, "c:\\foo").getParentFile().getName()).isEqualTo("c:");
+		assertThat(new File(base, "c:\\foo").getParentFile().getParentFile()).isEqualTo(base);
+		assertThat(new File(base, "\\\\sys\\foo").getName()).isEqualTo("foo");
+		assertThat(new File(base, "\\\\sys\\foo").getParentFile()
+			.getName()).isEqualTo("sys");
+		assertThat(new File(base, "\\\\sys\\foo").getParentFile()
+			.getParentFile()).isEqualTo(base);
+	}
+
+	@DisabledOnOs(WINDOWS)
+	@Test
+	public void testOnOtherOnly() {
+		File base = new File("base").getAbsoluteFile();
+		assertThat(new File(base, "c:\\foo").getName()).isEqualTo("c:\\foo");
+		assertThat(new File(base, "c:\\foo").getParentFile()).isEqualTo(base);
+		assertThat(new File(base, "\\\\sys\\foo").getName()).isEqualTo("\\\\sys\\foo");
+		assertThat(new File(base, "\\\\sys\\foo").getParentFile()).isEqualTo(base);
 	}
 
 	@Test
@@ -594,11 +616,6 @@ public class IOTest {
 			.getName()).contains("LPT1");
 		assertThat(os.getBasedFile(base, "bar/COM²²")
 			.getName()).contains("COM²²");
-
-		// not that a backslash cannot be tested since on Windows, the Path.of
-		// will translate them to segments. On other machines, the backslashes
-		// would not be removed. We can test with normal slashes here since
-		// windows nowadays treats them as segment separators.
 	}
 
 	private String except(RunnableWithException run) {

--- a/aQute.libg/test/aQute/lib/io/IOTest.java
+++ b/aQute.libg/test/aQute/lib/io/IOTest.java
@@ -558,7 +558,7 @@ public class IOTest {
 		assertThat("foo/💩../bar".contains("..")).isTrue();
 		assertThat("foo/.💩./bar".contains("..")).isFalse();
 		File base = new File("base").getAbsoluteFile();
-		assertThat(new File(base, "").getParentFile()).isEqualTo(base);
+		assertThat(new File(base, "")).isEqualTo(base);
 		assertThat(new File(base, "./").getParentFile()).isEqualTo(base);
 		assertThat(new File(base, "/bar").getParentFile()).isEqualTo(base);
 		assertThat(IO.getBasedFile(base, "")).isEqualTo(base);

--- a/aQute.libg/test/aQute/lib/io/IOTest.java
+++ b/aQute.libg/test/aQute/lib/io/IOTest.java
@@ -558,11 +558,12 @@ public class IOTest {
 		assertThat("foo/💩../bar".contains("..")).isTrue();
 		assertThat("foo/.💩./bar".contains("..")).isFalse();
 		File base = new File("base").getAbsoluteFile();
+		assertThat(new File(base, "").getParentFile()).isEqualTo(base);
+		assertThat(new File(base, "./").getParentFile()).isEqualTo(base);
 		assertThat(new File(base, "/bar").getParentFile()).isEqualTo(base);
 		assertThat(IO.getBasedFile(base, "")).isEqualTo(base);
 		assertThat(IO.getBasedFile(base, "foo/../bar")).isEqualTo(new File(base, "bar"));
 		assertThat(IO.getBasedFile(base, ".💩.")).isEqualTo(new File(base, ".💩."));
-		assertThat(IO.getBasedFile(base, ".x\u0008.")).isEqualTo(new File(base, ".x\u0008."));
 		assertThat(except(() -> IO.getBasedFile(base, ".."))).contains("io.sub.up");
 		assertThat(except(() -> IO.getBasedFile(base, "bar/../../.."))).contains("io.sub.up");
 	}

--- a/aQute.libg/test/aQute/lib/io/IOTest.java
+++ b/aQute.libg/test/aQute/lib/io/IOTest.java
@@ -561,6 +561,8 @@ public class IOTest {
 		assertThat(new File(base, "")).isEqualTo(base);
 		assertThat(new File(base, "./").getParentFile()).isEqualTo(base);
 		assertThat(new File(base, "/bar").getParentFile()).isEqualTo(base);
+		assertThat(IO.getBasedFile(base, "META-INF/versions/11/OSGI-INF/MANIFEST.MF")
+			.getName()).isEqualTo("MANIFEST.MF");
 		assertThat(IO.getBasedFile(base, "")).isEqualTo(base);
 		assertThat(IO.getBasedFile(base, "foo/../bar")).isEqualTo(new File(base, "bar"));
 		assertThat(IO.getBasedFile(base, ".💩.")).isEqualTo(new File(base, ".💩."));
@@ -578,6 +580,8 @@ public class IOTest {
 	public void testWindows() throws IOException {
 		Windows os = new Windows();
 		File base = new File("base").getAbsoluteFile();
+		assertThat(os.getBasedFile(base, "META-INF/versions/11/OSGI-INF/MANIFEST.MF")
+			.getName()).isEqualTo("MANIFEST.MF");
 		assertThat(except(() -> os.getBasedFile(base, "bar/LPT1"))).contains("io.win.sub.invalid");
 		assertThat(except(() -> os.getBasedFile(base, "bar/<bla"))).contains("io.win.sub.invalid");
 		assertThat(except(() -> os.getBasedFile(base, "bar/>bla/foo"))).contains("io.win.sub.invalid");

--- a/aQute.libg/test/aQute/lib/io/IOTest.java
+++ b/aQute.libg/test/aQute/lib/io/IOTest.java
@@ -10,6 +10,7 @@ import static org.junit.jupiter.api.Assertions.fail;
 import static org.junit.jupiter.api.condition.OS.WINDOWS;
 
 import java.io.File;
+import java.io.IOException;
 import java.io.InputStream;
 import java.io.Writer;
 import java.nio.ByteBuffer;
@@ -25,6 +26,7 @@ import org.junit.jupiter.api.condition.DisabledOnOs;
 import org.junit.jupiter.api.condition.EnabledOnOs;
 
 import aQute.bnd.exceptions.BiFunctionWithException;
+import aQute.bnd.exceptions.RunnableWithException;
 import aQute.bnd.test.jupiter.InjectTemporaryDirectory;
 import aQute.lib.io.IO.EnvironmentCalculator;
 
@@ -40,7 +42,6 @@ public class IOTest {
 				return map.getOrDefault(key, System.getenv(key));
 			}
 		};
-
 
 		assertEquals(new File(System.getProperty("user.home")), IO.home);
 
@@ -551,4 +552,57 @@ public class IOTest {
 
 		assertThat(Files.getPosixFilePermissions(target.toPath())).contains(PosixFilePermission.OWNER_EXECUTE);
 	}
+
+	@Test
+	public void testGetBasedFile() throws IOException {
+		assertThat("foo/💩../bar".contains("..")).isTrue();
+		assertThat("foo/.💩./bar".contains("..")).isFalse();
+		File base = new File("base").getAbsoluteFile();
+		assertThat(new File(base, "/bar").getParentFile()).isEqualTo(base);
+		assertThat(IO.getBasedFile(base, "")).isEqualTo(base);
+		assertThat(IO.getBasedFile(base, "foo/../bar")).isEqualTo(new File(base, "bar"));
+		assertThat(IO.getBasedFile(base, ".💩.")).isEqualTo(new File(base, ".💩."));
+		assertThat(IO.getBasedFile(base, ".x\u0008.")).isEqualTo(new File(base, ".x\u0008."));
+		assertThat(except(() -> IO.getBasedFile(base, ".."))).contains("io.sub.up");
+		assertThat(except(() -> IO.getBasedFile(base, "bar/../../.."))).contains("io.sub.up");
+	}
+
+	@Test
+	public void testCoder() {
+		String a = "💩..";
+		"..".indexOf(a);
+	}
+
+	@Test
+	public void testWindows() throws IOException {
+		Windows os = new Windows();
+		File base = new File("base").getAbsoluteFile();
+		assertThat(except(() -> os.getBasedFile(base, "bar/LPT1"))).contains("io.win.sub.invalid");
+		assertThat(except(() -> os.getBasedFile(base, "bar/<bla"))).contains("io.win.sub.invalid");
+		assertThat(except(() -> os.getBasedFile(base, "bar/>bla/foo"))).contains("io.win.sub.invalid");
+		assertThat(except(() -> os.getBasedFile(base, "bar/bla\nfoo"))).contains("io.win.sub.invalid");
+		assertThat(except(() -> os.getBasedFile(base, "bar/bla:foo"))).contains("io.win.sub.invalid");
+		assertThat(except(() -> os.getBasedFile(base, "bar/COM²"))).contains("io.win.sub.invalid");
+		assertThat(except(() -> os.getBasedFile(base, "bar/NUL"))).contains("io.win.sub.invalid");
+		assertThat(except(() -> os.getBasedFile(base, "bar/nul"))).contains("io.win.sub.invalid");
+		assertThat(os.getBasedFile(base, "bar/xLPT1")
+			.getName()).contains("LPT1");
+		assertThat(os.getBasedFile(base, "bar/COM²²")
+			.getName()).contains("COM²²");
+
+		// not that a backslash cannot be tested since on Windows, the Path.of
+		// will translate them to segments. On other machines, the backslashes
+		// would not be removed. We can test with normal slashes here since
+		// windows nowadays treats them as segment separators.
+	}
+
+	private String except(RunnableWithException run) {
+		try {
+			run.run();
+			return null;
+		} catch (Exception e) {
+			return e.toString();
+		}
+	}
+
 }

